mava/core_jax.py

```diff
@@ -60,18 +60,41 @@ def launch(self) -> None:
         """Run the system."""
 
 
-class SystemBuilder(abc.ABC):
-    """Abstract system builder."""
+class BaseSystemComponent(abc.ABC):
+    """Base class for core components: builder, executor, trainer and dataserver"""
 
-    def __init__(
-        self,
-    ) -> None:
-        """System building init"""
+    def __init__(self) -> None:
+        """Core component init"""
+        self.store: SimpleNamespace = SimpleNamespace()
+        self.callbacks: List
+
+    def has(self, component: Type, subtypes: bool = False) -> bool:
+        """Checks if a component of the given type is in the system.
+
+        Args:
+            component: type of component for which to check.
+            subtypes: whether or not to check for subtypes of the given component.
 
-        # Simple namespace for assigning system builder attributes dynamically
-        self.store = SimpleNamespace()
+        Returns:
+            True if the system has a component of this type.
+        """
+        if subtypes:  # checks for type of component and its subtypes
+            return any(
```

Comment:
Feel free to ignore this. But can we use something like `check_fn = instance if subtypes else lambda x, y: type(x)==y` to simplify the code a bit?

Reply:
Thought about it a bit more and I think @ulricharmel is right, if we just have `isinstance` then we should be fine, you just need to pass the most specific component possible